File: android/app/src/main/java/com/frenchnclc7/app/data/SpeechPlayer.kt
package com.frenchnclc7.app.data

import android.content.Context
import android.media.MediaPlayer
import java.io.File

/** Plays the French pronunciation returned by the text-to-speech function (an MP3). One at a time. */
class SpeechPlayer(private val context: Context) {
    private var player: MediaPlayer? = null

    fun play(audio: ByteArray) {
        stop()
        try {
            val file = File(context.cacheDir, "speech.mp3")
            file.writeBytes(audio)
            val p = MediaPlayer()
            p.setDataSource(file.path)
            p.setOnCompletionListener {
                it.release()
                if (player === it) player = null
            }
            p.prepare()
            p.start()
            player = p
        } catch (e: Exception) {
            // Audio is a nicety; the card still works without it.
            player = null
        }
    }

    fun stop() {
        player?.let {
            try { it.stop() } catch (e: IllegalStateException) { /* already stopped */ }
            it.release()
        }
        player = null
    }
}

File: ios/FrenchNCLC7/Store/SpeechPlayer.swift
import AVFoundation
import Foundation

/// Plays the French pronunciation returned by the text-to-speech function (an MP3). One at a time.
final class SpeechPlayer {
    private var player: AVAudioPlayer?

    func play(_ audio: Data) {
        stop()
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .spokenAudio)
            try AVAudioSession.sharedInstance().setActive(true)
            let p = try AVAudioPlayer(data: audio)
            p.prepareToPlay()
            p.play()
            player = p
        } catch {
            // Audio is a nicety; the card still works without it.
            player = nil
        }
    }

    func stop() {
        player?.stop()
        player = nil
    }
}
